### PR TITLE
Add log events when blocks are added/removed to/from the blockstore

### PR DESCRIPTION
**Pull request by MichaelMure**
This PR add log events when are added/removed to/from the blockstore. I need these events for the Arbore project to track block locality.

License: MIT
Signed-off-by: Michael Muré <mure.michael@gmail.com>

### CODE REVIEW

**Comment by whyrusleeping** [comment]
I would go ahead and make those actual `log.Event`s, with the pattern of:
```
defer log.EventBegin(....).Done()
```

This will let us also extract timing information

**Comment by MichaelMure** [comment]
If I understand correctly, you need a Context to use this function and there is not in this part of the code. I would then need to pass the Context around and it does propagate in various subsystem (blockstore, exchange, merkledag at least).

Is it worth it ?

**Comment by whyrusleeping** [comment]
@MichaelMure just use a `context.TODO()` or a `context.Background()` for now

**Comment by jbenet** [comment]
`context.TODO()` please

**Comment by MichaelMure** [comment]
Will do.

**Comment by MichaelMure** [comment]
Well, turns out that `log.EventBegin(....).Done()` does not output logs to stderr, so I still need the regular `log.Info`.

**Comment by jbenet** [comment]
@MichaelMure do not depend on the logs from stderr -- those are for debugging and liable to change at any moment. use the logs from the events HTTP endpoint (`https://localhost:5001/logs` i think). supporting other tools is what they're for. they're nicely formatted as ndjson for your convenience.

**Comment by jbenet** [comment]
would be good to have sharness tests for events that people will depend on. that way we wont break them. maybe could make
- [ ] sharness func to log events to a file, run a cmd, and then check file has a certain event.
- [ ] add sharness test for this and every event you depend on, so we dont break them.

**Comment by MichaelMure** [comment]
Ok, so I finally use ipfs log tail instead of stderr. Can we merge this ?

File: blocks/blockstore/write_cache.go [hunk 1]
```diff
@@ -50,6 +54,8 @@ func (w *writecache) PutMany(bs []*blocks.Block) error {
 	for _, b := range bs {
 		if _, ok := w.cache.Get(b.Key()); !ok {
 			good = append(good, b)
+			k := b.Key();
+			defer log.EventBegin(context.TODO(), "Added block ", &k).Done()
```
Comment (jbenet) [inline blocks/blockstore/write_cache.go]: this is an odd placement for this, @whyrusleeping thoughts?

**Comment by jbenet** [comment]
Ok this LGTM. 

weird: this failed:

- https://travis-ci.org/ipfs/go-ipfs/jobs/91023735#L4203-L4226

@whyrusleeping  o/

File: blocks/blockstore/write_cache.go (hunk 1 above)
Comment (whyrusleeping) [inline blocks/blockstore/write_cache.go]: while it seems weird, it makes sense to me. We are registering the block to be stored here, but before the defer fires, the actual put will happen: `return w.blockstore.PutMany(good)`. So it should work as expected.

File: blocks/blockstore/write_cache.go [hunk 2]
```diff
@@ -22,6 +22,7 @@ type writecache struct {
 }
 
 func (w *writecache) DeleteBlock(k key.Key) error {
+	defer log.EventBegin(context.TODO(), "Removed block ", &k).Done()
```
Comment (whyrusleeping) [inline blocks/blockstore/write_cache.go]: remove trailing spaces from the string contant, and change the event label to something like `writecache.BlockRemoved`

**Comment by whyrusleeping** [comment]
sorry it took so long to get to this, one comment, and you'll need to run `go fmt` over the code (i see some semicolons, C++ eh?)

Otherwise this LGTM, i'm always in favor of having more useful eventlogs like this

**Comment by whyrusleeping** [comment]
also rebasing on latest master would be great too, thanks!

File: blocks/blockstore/write_cache.go (hunk 2 above)
Comment (MichaelMure) [inline blocks/blockstore/write_cache.go]: IIRC, the trailing space was needed to separate the key and "Removed block" in the text logs (not json).

**Comment by MichaelMure** [comment]
Dammit, the semicolons ;-) C++ and PHP currently. 

I will rebase and fix later.

**Comment by MichaelMure** [comment]
@whyrusleeping Fixed and rebased.

**Comment by whyrusleeping** [comment]
@MichaelMure LGTM, sorry for the delay.

**Pull request merged**